src/main/java/ether/EtherClient.java

```diff
@@ -0,0 +1,719 @@
+        byte [] approveFunctionBytes=approve.encode(Settings.CHANNEL_ADDRESS, deposit);
+
+        String queryApproveGasString = "{\"method\":\"eth_estimateGas\"," +
+                "\"params\":[" +
+                "{" +
+                "\"from\":\""+senderAddress+"\"," +
+                "\"to\":\""+Settings.TOKEN_ADDRESS+"\"," +
+                "\"value\":\""+"0x"+new BigInteger("0",10).toString(16)+"\","+
+                "\"data\":\""+"0x" + new String(org.apache.commons.codec.binary.Hex.encodeHex(approveFunctionBytes))+"\"" +
+                "}" +
+                "]," +
+                "\"id\":42,\"jsonrpc\":\"2.0\"}";
+
+        if(Settings.DEBUG_INFO) {
+            System.out.println("The request string of queryApproveGasString is "+queryApproveGasString);
+        }
+
+        String approveGasEstimateStr;
+        try {
+            approveGasEstimateStr=(String)httpAgent.getHttpResponse(queryApproveGasString);
+        }catch (IOException e) {
+            if(Settings.DEBUG_INFO) {
+                System.out.println("Invoking function with given arguments is not allowed.");
+            }
+            throw e;
+        }
+
+        String senderNonce = getNonce(senderAddress, httpAgent).toString(); //TODO: check base, 10 or 16
+
+        if(Settings.DEBUG_INFO) {
+            System.out.println("The estimatedGas of approve is "+approveGasEstimateStr+".");
+            System.out.println("The nonce of "+senderAddress+" is "+senderNonce);
+        }
+
+        BigInteger approveGasEstimate = new BigInteger(approveGasEstimateStr); //TODO: check base, 10 or 16
+
+        if(approveGasEstimate.compareTo(Settings.GAS_LIMIT) > 0){
+            if(Settings.DEBUG_INFO) {
+                System.out.println("Estimeted GAS for approve transaction is too high. Probably some arguments are Invalid.");
+            }
+            throw new IllegalArgumentException("Estimeted GAS for approve transaction is too high. Probably some arguments are Invalid.");
+        }
+
+
+        String approveSendRawTransactionString = "{\"method\":\"eth_sendRawTransaction\",\"params\":[\""
+                + signedApproveTrans + "\"],\"id\":42,\"jsonrpc\":\"2.0\"}";
+
+        String transactionId;
+        try {
+            transactionId=(String)httpAgent.getHttpResponse(approveSendRawTransactionString);
+        }catch (IOException e) {
+            if(Settings.DEBUG_INFO) {
+                System.out.println("Fail to execute HTTP request.");
+            }
+            throw e;
+        }
+
+        if(transactionId.equals("")){
+            if(Settings.DEBUG_INFO) {
+                System.out.println("Fail to submit the approve transaction.");
+            }
+            return false;
+        }
+
+        if(Settings.DEBUG_INFO) {
+            System.out.println("Waiting for network to mine transactions ... ");
+        }
+
+        waitingForTransaction(transactionId, httpAgent);
```

Comment:
are we running the whole transaction stuff in a separate thread - just in case if it takes too much time to mine?

Reply:
You are right Sri, each transaction is executed on a different Thread.

//execute on a separate Thread
executor.execute(() -> processTransaction(event));